src/providers/openstack/configdrive.rs

```diff
@@ -0,0 +1,166 @@
+        let path = Path::new("/media/ConfigDrive/openstack/latest/");
+        if path.exists() {
+            return Ok(OpenstackConfigDrive {
+                temp_dir: None,
+                drive_path: PathBuf::from("/media/ConfigDrive/"),
+            });
+        }
+
+        // Otherwise, try and mount with the label.
+        let target = tempfile::Builder::new()
+            .prefix("afterburn-")
+            .tempdir()
+            .chain_err(|| "failed to create temporary directory")?;
+        crate::util::mount_ro(
+            &Path::new("/dev/disk/by-label/").join(CONFIG_DRIVE_LABEL),
+            target.path(),
+            "iso9660",
+            3,
+        )?;
+
+        let cd = OpenstackConfigDrive {
+            drive_path: target.path().to_owned(),
+            temp_dir: Some(target),
+        };
+        Ok(cd)
+    }
+
+    /// Return the path to the metadata directory.
+    fn metadata_dir(&self, platform: &str) -> PathBuf {
+        self.drive_path.clone().join(platform).join("latest")
+    }
+
+    // The metadata is stored as key:value pair in ec2/meta-data.json file
+    fn fetch_value(&self, key: &str) -> Result<Option<String>> {
+        let filename = self.metadata_dir("ec2").join("meta-data.json");
+
+        if !filename.exists() {
+            return Ok(None);
+        }
+
+        let mut file =
+            File::open(&filename).chain_err(|| format!("failed to open file '{:?}'", filename))?;
+
+        let mut contents = String::new();
+        file.read_to_string(&mut contents)
+            .chain_err(|| format!("failed to read from file '{:?}'", filename))?;
+        let contents_json: serde_json::Value = serde_json::from_str(&contents)?;
+
+        if let Some(value) = contents_json[key].as_str() {
+            Ok(Some(value.to_owned()))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn fetch_publickeys(&self) -> Result<Vec<PublicKey>> {
+        let filename = self.metadata_dir("openstack").join("meta_data.json");
+        let mut file =
+            File::open(&filename).chain_err(|| format!("failed to open file '{:?}'", filename))?;
+
+        let mut contents = String::new();
+        file.read_to_string(&mut contents)
+            .chain_err(|| format!("failed to read from file '{:?}'", filename))?;
+        let contents_json: serde_json::Value = serde_json::from_str(&contents)?;
+
+        if let Some(public_keys_map) = contents_json["public_keys"].as_object() {
+            let public_keys_vec: Vec<&std::string::String> = public_keys_map.keys().collect();
+            let mut out = vec![];
+            for key in public_keys_vec {
+                let key = PublicKey::parse(&key)?;
+                out.push(key);
+            }
+            Ok(out)
+        } else {
+            Ok(vec![])
+        }
+    }
+}
+
+impl MetadataProvider for OpenstackConfigDrive {
+    fn attributes(&self) -> Result<HashMap<String, String>> {
```

Comment:
It would be good to have some unit-tests on this (see the ibmcloud-* providers for examples).

Reply:
Added unit tests for attributes and ssh_keys.